Fix building for host architecture by default

RoboTux opened the pull request:
By default Tensorflow build aims to only build what is relevant for the host targets. However, due to some missing guards here and there code for other architecture may end up being built. For instance, when building tensorflow/compiler/mlir:tf-opt on AArch64 code specific to x86 also ends up being built. This commit fixes that (aka issue #59595).

RoboTux committed:
Fix building for host architecture by default

By default Tensorflow build aims to only build what is relevant for the
host targets. However, due to some missing guards here and there code
for other architecture may end up being built. For instance, when
building tensorflow/compiler/mlir:tf-opt on AArch64 code specific to x86
also ends up being built. This commit fixes that (aka issue #59595).

cheshire commented:
@tpopp could you take a look?

tpopp commented on tensorflow/compiler/xla/runtime/BUILD:
```diff
@@ -257,26 +257,15 @@ cc_library(
-            "@llvm-project//llvm:AArch64CodeGen",  # fixdeps: keep
-        ],
-        "//tensorflow/tsl:linux_ppc64le": [
-            "@llvm-project//llvm:PowerPCAsmParser",  # fixdeps: keep
```
Is it purposeful that some of these dependencies were fully removed?

RoboTux replied:
No, I must have made a mistake in my rebase. Let me correct that.

tpopp reviewed (dismissed):
Thanks for doing this. It is not feasible for me to understand if all required dependencies are supported with this change, so I'll run the test suite and base any feedback on failing CI or all passing CI.

tpopp commented:
Checking, should I review this again or not (it is marked as draft)?

RoboTux commented:
> Hi @tpopp! @RoboTux is ooo until April 11th. He asked me to let you know he wanted to have a look at the CI failures before taking it out of draft. Cheers and have a nice weekend!

I'm slowly catching up but the problem seems to be that tensorflow/python/tools/saved_model_cli_test.py assumes both x86 and aarch64 are always available. I'll see how to remove that expectation on Monday, any suggestion welcome though.

tpopp commented:
I'll share it with others in the US in hopes that someone else has thoughts but won't be able to look earlier than Monday or Tuesday myself.

RoboTux committed:
Fix python tests and reorder aarch32/aarch64

RoboTux commented:
FYI: I have a revised patch that I'm going to push shortly (probably tomorrow).

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux committed:
Fix AArch64 CI build failure

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

tpopp commented:
Hey, sorry, I just came back from vacation. I'm going to see if I can find a more appropriate person to review this and will try to, in parallel, make sure everything passes, so there are no unexpected surprises later.

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

ddunl commented:
To me it seems like @jakeharmon8 and @MichaelHudgins might be the ones who should look at this?

RoboTux commented:
Failure seems unrelated:
"ModuleNotFoundError: No module named 'tensorflow.compiler.mlir.stablehlo'"

It also passed CI before this last merge.

RoboTux commented:
> Failure seems unrelated: "ModuleNotFoundError: No module named 'tensorflow.compiler.mlir.stablehlo'"
> 
> It also passed CI before this last merge.

Same error on master as well: https://github.com/tensorflow/tensorflow/actions/runs/5010911951/jobs/8981212479

elfringham commented:
@RoboTux That failure was introduced with https://github.com/tensorflow/tensorflow/commit/c09dcc2c7d07a84cdb9f9b5342d7803f5dbc50b3

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

MichaelHudgins commented:
@vam-google or @jakeharmon8 could you PTAL?

RoboTux committed:
Merge remote-tracking branch 'origin/master' into new

RoboTux commented:
Ping?

RoboTux committed:
Fix include failure in compilation_pipeline_cpu.cc

RoboTux committed:
Merge remote-tracking branch 'origin/master' into new

RoboTux commented:
I'm working on fixing remaining issues.

RoboTux committed:
Add missing guards for arch-specific code

RoboTux committed:
Fix Windows build

RoboTux committed:
Fix codestyle

RoboTux committed:
Act test API change

RoboTux committed:
Merge remote-tracking branch 'origin/master' into new

RoboTux commented:
The latest push should fix the API and codestyle issues. Note: this does indeed introduce a new function in the test API and follows a similar pattern to ROCm and cuda. Let me know if it isn't fine.

RoboTux committed:
Remove custom copts in passes library

RoboTux commented:
All remaining issues seem unrelated to this change:
- Arm CI fails to find the pip package it created (ls: cannot access '/tensorflow/whl/-*310*310*.whl': No such file or directory)
- AMD ROCm fails to find clang
- MacOS has 16 tests timing out

RoboTux commented:
Ping?

MichaelHudgins commented:
@jakeharmon8 can you rereview? 
@RoboTux can you please rebase?

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux commented:
Windows build failure looks unrelated:

```
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2065: 'not': undeclared identifier
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2146: syntax error: missing ')' before identifier 'scalables'
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(118): error C2059: syntax error: ')'
external/llvm-project/mlir/lib/Interfaces/ViewLikeInterface.cpp(119): error C2146: syntax error: missing ';' before identifier 'printer'
```

RoboTux commented:
MacOS build seems to fail for performance reasons if I read this correctly:
```
//tensorflow/python/kernel_tests/linalg/sparse:csr_sparse_matrix_sparse_mat_mul_grad_test_cpu FAILED in 150 out of 150 in 1.2s
  Stats over 150 runs: max = 1.2s, min = 0.5s, avg = 0.8s, dev = 0.1s
```

Since this code only changes the build system that must be unrelated.

As to ARM CI there is one test where TF crashes which I'm not sure about and one x86 dialect loading failure which is most likely related to this diff. I'll have a look now

```
FAIL: //tensorflow/compiler/xla/mlir/math/transforms/tests:math_optimization.mlir.test
(...)
LLVM ERROR: Building op `x86vector.avx.rsqrt` but it isn't registered in this MLIRContext: the dialect may not be loaded or this operation isn't registered by the dialect.
```

elfringham commented:
//tensorflow/dtensor/mlir/tests:spmd_expansion.mlir.test is known as flaky, see https://github.com/tensorflow/tensorflow/issues/61164

RoboTux committed:
Guard x86-specific math opt testcases

Move x86-specific math optimization tests to a separate file to be able
to guard it on x86 target support being available.

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

freddan80 commented:
fyi - @RoboTux is on leave until 26th.

johnnkp commented:
I can confirm that this PR works for Windows CUDA build. The remaining are resolving conflicts and merge quickly before new merge conflicts happens.

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

hawkinsp commented on tensorflow/tsl/platform/default/build_config.bzl:
```diff
@@ -862,14 +862,44 @@ def tf_google_mobile_srcs_no_runtime():
 def if_llvm_aarch64_available(then, otherwise = []):
-    return then
+    return select({
+        clean_dep("//tensorflow/tsl:linux_aarch64"): then,
```
Aren't we missing Mac ARM here (and Mac x86 below as well)?

RoboTux replied:
Indeed, missed that sorry. I'll upload a new version.

RoboTux committed:
Fix dependencies for MacOS.

RoboTux commented:
Ping?

jakeharmon8 commented on tensorflow/python/platform/test.py:
```diff
@@ -192,3 +192,25 @@ def is_built_with_xla():
+def is_cpu_target_available(target):
+  """Returns whether TensorFlow was built for the given CPU target.
+
+  This method should only be used in tests written with `tf.test.TestCase`. A
```
Add an "Args" section here to appease our linter (e.g. see line 82)

jakeharmon8 reviewed (changes requested):
Add a note about the API addition in RELEASE.md (an internal check requires it when making API changes)

jakeharmon8 commented:
There are two more blockers that I'm working on:

1) Internal technical issues that I should have fixed soon
2) Changes to the API normally require an RFC+Review (see [here](https://github.com/tensorflow/community/blob/master/rfcs/yyyymmdd-rfc-template.md)). I'm looking into what the process is for external contributors, and whether or not there's a shorter process for small changes like this. If you'd like to avoid that red tape, you could refactor this to remove the API changes.

RoboTux committed:
Document arguments to is_cpu_target_available()

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux commented:
> There are two more blockers that I'm working on:
> 
>     1. Internal technical issues that I should have fixed soon
> 
>     2. Changes to the API normally require an RFC+Review (see [here](https://github.com/tensorflow/community/blob/master/rfcs/yyyymmdd-rfc-template.md)). I'm looking into what the process is for external contributors, and whether or not there's a shorter process for small changes like this. If you'd like to avoid that red tape, you could refactor this to remove the API changes.

I'm not sure how I could remove the API change. I need a way to skip a target CPU specific test if the support for that target is not built in so need a way to query that.

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

jakeharmon8 commented:
The conflicts are that many of these files moved - I think I can just merge the internal version (since they didn't move internally). Let me try getting all the approvals again tomorrow and hopefully I can ship this.

jakeharmon8 commented:
When these files moved, copybara deleted the internal CL and created a new one, which means I have to start over on all of the internal approvals. Sorry for the comical amount of red tape - I should have this merged any day now.

ebrevdo commented:
Sorry for the late join on this PR.  The current behavior of XLA AOT generation is to support x-compiling across multiple architectures.  This is used, by example, by the `saved_model_cli` tool.  It's OK to disable when building TF in general, but for the PyPi build we should allow the existing behavior.  Can you ensure that this PR continues to build TF PyPi packages that allow XLA AOT cross-compilation across all existing known architectures?

RoboTux commented:
> Sorry for the late join on this PR. The current behavior of XLA AOT generation is to support x-compiling across multiple architectures. This is used, by example, by the `saved_model_cli` tool. It's OK to disable when building TF in general, but for the PyPi build we should allow the existing behavior. Can you ensure that this PR continues to build TF PyPi packages that allow XLA AOT cross-compilation across all existing known architectures?
> 
> I _think_ the changes would have to be made in the files in: ci/official/envs

I'll double check.

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux committed:
Enable all targets in CI builds

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

MichaelHudgins commented on ci/official/envs/continuous_linux_x86_cpu_py310:
You can update in only a single place within the root .bazelrc instead of in each env if this is needed in either all release configs or all linux release configs by adding to release_base or release_cpu_linux respectively.

MichaelHudgins replied:
Also side note, those envs are currently not used by the prod jobs.  We will be switching to them in the near future though.  Adding to the root .bazelrc will catch both the older and newer ci jobs.

RoboTux committed:
Pull out enabling cross compiling for CI in .bazelrc

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

ebrevdo commented:
We should get this in!

On Fri, Oct 20, 2023, 6:47 PM github-actions[bot] ***@***.***>
wrote:

> This PR is stale because it has been open for 14 days with no activity. It
> will be closed if no further activity occurs. Thank you.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/60161#issuecomment-1773595281>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANWFG2TQJTZGDJEX7ML2R3YAMSSPAVCNFSM6AAAAAAWLVWTACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZTGU4TKMRYGE>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

jakeharmon8 committed:
Merge branch 'master' into fix_building_for_host_arch

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

RoboTux committed:
Merge remote-tracking branch 'origin/master' into fix_building_for_host_arch

RoboTux commented:
> Hi @RoboTux Can you please resolve conflicts? Thank you!

Done.

Note 1: I'm likely to soon have less time to push this patch forward and my have to drop it if it doesn't get approved soon.
Note 2: The patch contains changes to third_party/xla and third_party/xla/third_party/tsl. I'm looking for a general approval before submitting these changes separately to the appropriate repos.

Best regards,
Thomas

RoboTux commented:
Ping @jakeharmon8 ?

RoboTux committed:
Fix bazel codestyle

RoboTux commented:
The PyLint issue is unrelated. It is due to the import of gen_sparse_ops added by https://github.com/tensorflow/tensorflow/commit/d7030ffa0e0a90378f3a4efc099cc50fd07c0577

jakeharmon8 commented:
The Kafkaesque nightmare continues - two internal approvals left.

RoboTux commented:
> The Kafkaesque nightmare continues - two internal approvals left.

Thanks for the update and for pushing this. Much appreciated.

gbaned commented:
Hi @RoboTux Can you please resolve conflicts? Thank you!

jakeharmon8 commented:
Don't worry about rebasing - I'm still managing this internally. It had broken some internal tests concerning LLVM and Arm64, but I think I've almost got that working. I'll jump back on this today.

ddunl commented:
It seems like this landed, so we can close. @jakeharmon8 ?

RoboTux commented:
> It seems like this landed, so we can close. @jakeharmon8 ?

If it indeed has thanks Jake!

jakeharmon8 commented:
Yup, it finally landed. Sorry for the delay - this PR really was the perfect storm of creating as much internal bureaucracy as possible!
